pydicom/fileset.py

```diff
@@ -0,0 +1,2861 @@
+            The keyword or tag of the element to search for.
+        instances : list of pydicom.fileset.FileInstance, optional
+            Search within the given instances. If not used then all available
+            instances will be searched.
+        load : bool, optional
+            If ``True``, then load the SOP Instances belonging to the
+            File-set and perform the search against their available elements.
+            Otherwise (default) search only the elements available in the
+            corresponding directory records (more efficient, but only a limited
+            number of elements are available).
+
+        Returns
+        -------
+        list of object
+            A list of value(s) for the element available in the instances.
+        """
+        has_element = False
+        results = []
+        instances = instances or iter(self)
+        for instance in instances:
+            if load:
+                instance = instance.load()
+
+            if element not in instance:
+                continue
+
+            has_element = True
+            val = instance[element].value
+            # Not very efficient, but we can't use set
+            if val not in results:
+                results.append(val)
+
+        if not load and not has_element:
+            warnings.warn(
+                "None of the records in the DICOMDIR dataset contain "
+                "the query element, consider using the 'load' parameter "
+                "to expand the search to the corresponding SOP instances"
+            )
+
+        return results
+
+    @property
+    def ID(self) -> Union[str, None]:
+        """Return the *File-set ID* (if available) or ``None``."""
+        return self._id
+
+    @ID.setter
+    def ID(self, val: Union[str, None]) -> None:
+        """Set the File-set ID.
+
+        Parameters
+        ----------
+        val : str or None
+            The value to use for the DICOMDIR's (0004,1130) *File-set ID*.
+
+        Raises
+        ------
+        ValueError
+            If `val` is greater than 16 characters long.
+        """
+        if val == self._id:
+            return
+
+        if val is None or 0 <= len(val) <= 16:
+            self._id = val
+            if self._ds:
+                self._ds.FileSetID = val
+            self._stage['^'] = True
+        else:
+            raise ValueError(
+                "The maximum length of the 'File-set ID' is 16 characters"
+            )
+
+    @property
+    def is_staged(self) -> bool:
+        """Return ``True`` if the File-set is new or has changes staged."""
+        is_staged = [
+            self._stage['+'],
+            self._stage['-'],
+            self._stage['~'],
+            self._stage['^'],
+        ]
+        return any(is_staged)
+
+    def __iter__(self) -> FileInstance:
+        """Yield :class:`~pydicom.fileset.FileInstance` from the File-set."""
+        yield from self._instances[:]
+
```

Comment:
Does the `[:]` list copy change anything?  Still get the same references within the list

Reply:
I had failing tests with:
```python
for instance in fs:
    fs.remove(instance)
```
Because it was removing from the list it was iterating from it'd skip every second item.

Reply:
Right - makes sense.